File: src/types/flashcard.ts
export type PartOfSpeech = 
  | 'noun' 
  | 'verb' 
  | 'adjective' 
  | 'adverb' 
  | 'idiom' 
  | 'phrasal verb' 
  | 'preposition' 
  | 'conjunction' 
  | 'grammar structure'
  | 'other';

export interface Flashcard {
  id: string;
  term: string;                      // Từ vựng hoặc Cụm từ/Cấu trúc
  phonetic?: string;                  // Phiên âm IPA (VD: /prəˈfɪʃ.ənt/)
  partOfSpeech?: string;              // Từ loại (VD: adj, v, n, idiom...)
  definition: string;                // Định nghĩa / Nghĩa tiếng Việt
  example?: string;                   // Câu ví dụ minh hoạ
  exampleTranslation?: string;        // Dịch nghĩa câu ví dụ
  grammarPattern?: string;            // Cấu trúc/Mẫu ngữ pháp (VD: proficient in/at sth)
  notes?: string;                     // Ghi chú, mẹo nhớ (mnemonics), từ đồng nghĩa/trái nghĩa
  tags?: string[];                    // Nhãn phân loại (C1, IELTS, Topic, etc.)
  starred?: boolean;                  // Đánh dấu yêu thích / cần chú ý

  // Dữ liệu thuật toán SRS (Spaced Repetition - SM2)
  repetition: number;                 // Số lần ôn tập thành công liên tiếp
  interval: number;                   // Khoảng cách ngày ôn tập tiếp theo
  easeFactor: number;                 // Hệ số độ khó (mặc định 2.5)
  dueDate: string;                    // Ngày cần ôn tập tiếp theo (ISO string)
  lastReviewed?: string;              // Ngày ôn tập gần nhất
  reviewCount: number;                // Tổng số lần đã học
  lapses: number;                     // Số lần quên (nhấn Again)
}

export interface Deck {
  id: string;
  title: string;
  description: string;
  language: string;                   // 'en-US' | 'en-GB' | 'ja-JP' | 'zh-CN' | 'ko-KR' | 'fr-FR' | etc.
  targetLanguage: string;             // 'vi-VN'
  category?: string;
  tags: string[];
  cards: Flashcard[];
  createdAt: string;
  updatedAt: string;
  color?: string;                     // Accent color theme cho bộ thẻ
}

export type StudyMode = 'flashcards' | 'srs' | 'write' | 'match' | 'test';

export type SRSRating = 1 | 2 | 3 | 4; // 1: Again (Quên), 2: Hard (Khó), 3: Good (Tốt), 4: Easy (Dễ)

export interface ColumnMapping {
  term: string;
  phonetic: string;
  partOfSpeech: string;
  definition: string;
  example: string;
  exampleTranslation: string;
  grammarPattern: string;
  notes: string;
  tags: string;
}

export interface TestResult {
  total: number;
  correct: number;
  incorrect: number;
  percentage: number;
  timeSpentSeconds: number;
  answers: {
    cardId: string;
    term: string;
    userAnswer: string;
    correctAnswer: string;
    isCorrect: boolean;
    type: 'multiple-choice' | 'written' | 'true-false';
  }[];
}

File: pure-macos/Sources/Models/Flashcard.swift
import Foundation

public struct Flashcard: Identifiable, Codable, Equatable, Hashable {
    public var id: String
    public var term: String
    public var phonetic: String?
    public var partOfSpeech: String?
    public var definition: String
    public var example: String?
    public var exampleTranslation: String?
    public var grammarPattern: String?
    public var notes: String?
    public var tags: [String]
    public var starred: Bool
    
    // Spaced Repetition (SuperMemo SM-2) Data
    public var repetition: Int
    public var interval: Int
    public var easeFactor: Double
    public var dueDate: Date
    public var lastReviewed: Date?
    public var reviewCount: Int
    public var lapses: Int
    
    public init(
        id: String = UUID().uuidString,
        term: String,
        phonetic: String? = nil,
        partOfSpeech: String? = nil,
        definition: String,
        example: String? = nil,
        exampleTranslation: String? = nil,
        grammarPattern: String? = nil,
        notes: String? = nil,
        tags: [String] = [],
        starred: Bool = false,
        repetition: Int = 0,
        interval: Int = 1,
        easeFactor: Double = 2.5,
        dueDate: Date = Date(),
        lastReviewed: Date? = nil,
        reviewCount: Int = 0,
        lapses: Int = 0
    ) {
        self.id = id
        self.term = term
        self.phonetic = phonetic
        self.partOfSpeech = partOfSpeech
        self.definition = definition
        self.example = example
        self.exampleTranslation = exampleTranslation
        self.grammarPattern = grammarPattern
        self.notes = notes
        self.tags = tags
        self.starred = starred
        self.repetition = repetition
        self.interval = interval
        self.easeFactor = easeFactor
        self.dueDate = dueDate
        self.lastReviewed = lastReviewed
        self.reviewCount = reviewCount
        self.lapses = lapses
    }
    
    public var isDue: Bool {
        return dueDate <= Date()
    }
}
